### PR TITLE
Add LICENSE test to presubmit checks

### LINKED ISSUE #28368: Check LICENSE in flutter/flutter
Some dart files are missing LICENSE headers. We should add a presubmit test to check that. See also https://github.com/flutter/flutter/pull/27660#discussion_r259502544

### PR DESCRIPTION
**Pull request by liyuqian**
## Description

Also update the existing dart files with missing licenses.

Without the fix, we'll emit the following error message
```
License headers cannot be found at the beginning of the following files.

/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/lib/src/animation/tween_sequence.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/material/raw_material_button_test.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/async_lifecycle_test.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/sliver_constraints_test.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/app_test.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/test_border.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/physical_model_test.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter/test/widgets/inherited_model.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter_tools/lib/src/base/user_messages.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter_tools/test/src/pubspec_schema.dart
/usr/local/google/home/liyuqian/flutter/flutter/packages/flutter_tools/test/ios/simulators_test.dart
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

## Related Issues

Fixes https://github.com/flutter/flutter/issues/28368

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: dev/bots/analyze.dart [hunk 1]
```diff
@@ -472,6 +473,36 @@ Future<void> _verifyNoBadImportsInFlutterTools(String workingDirectory) async {
...
+final RegExp _licensePattern = RegExp(r'''LICENSE file''');
+
+Future<void> _verifyNoMissingLicense(String workingDirectory) async {
+  final List<String> errors = <String>[];
+  for (FileSystemEntity entity in Directory(path.join(workingDirectory, 'packages'))
+      .listSync(recursive: true)
+      .where((FileSystemEntity entity) => entity is File && path.extension(entity.path) == '.dart')) {
+    final File file = entity;
+    bool hasLicense = false;
+    for (String line in file.readAsLinesSync()) {
```
Comment (goderbauer) [inline dev/bots/analyze.dart]: Don't we want to enforce that the license header is at the beginning of the file?

Comment (liyuqian) [reply dev/bots/analyze.dart]: I guess so. Is it safe to assume that `LICENSE file` should appear in the top 10 lines of the file?

Comment (goderbauer) [reply dev/bots/analyze.dart]: Would it be problematic to just expect that the first line of the file starts with "Copyright \d{4}"?

Comment (liyuqian) [reply dev/bots/analyze.dart]: Good idea. Done 😄

File: dev/bots/analyze.dart [hunk 2]
```diff
@@ -472,6 +473,32 @@ Future<void> _verifyNoBadImportsInFlutterTools(String workingDirectory) async {
   }
 }
 
+final RegExp _licensePattern = RegExp(r'''Copyright \d{4}''');
```
Comment (goderbauer) [inline dev/bots/analyze.dart]: Can this just use single quotes `'` instead of `'''`?

File: dev/bots/analyze.dart [hunk 3]
```diff
@@ -472,6 +473,32 @@ Future<void> _verifyNoBadImportsInFlutterTools(String workingDirectory) async {
...
+Future<void> _verifyNoMissingLicense(String workingDirectory) async {
+  final List<String> errors = <String>[];
+  for (FileSystemEntity entity in Directory(path.join(workingDirectory, 'packages'))
+      .listSync(recursive: true)
+      .where((FileSystemEntity entity) => entity is File && path.extension(entity.path) == '.dart')) {
+    final File file = entity;
+    bool hasLicense = false;
+    final List<String> lines = file.readAsLinesSync();
+    if (lines.isNotEmpty)
+      hasLicense = _licensePattern.firstMatch(lines.first) != null;
```
Comment (goderbauer) [inline dev/bots/analyze.dart]: Could this just be `lines.first.startsWith(_licensePattern)` with `_licensePattern = RegExp(r'// Copyright \d{4}''');`

**Review by goderbauer — APPROVED** [review]
LGTM

File: dev/bots/analyze.dart (hunk 2 above)
Comment (liyuqian) [reply dev/bots/analyze.dart]: Yes.

File: dev/bots/analyze.dart (hunk 3 above)
Comment (liyuqian) [reply dev/bots/analyze.dart]: Done.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/test/material/raw_material_button_test.dart [hunk 4]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/material/raw_material_button_test.dart]: Technically, I think the year to use is the one in which this file was originally authored, not the day the copyright notice was added.  You can find this out in the git log for the file.

File: packages/flutter/lib/src/animation/tween_sequence.dart [hunk 5]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/animation/tween_sequence.dart]: Technically, I think the year to use is the one in which this file was originally authored, not the day the copyright notice was added. You can find this out in the git log for the file.

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/animation/tween_sequence.dart (hunk 5 above)
Comment (liyuqian) [reply packages/flutter/lib/src/animation/tween_sequence.dart]: Changed to 2018 according to the earliest commit in git log.

File: packages/flutter/test/material/raw_material_button_test.dart (hunk 4 above)
Comment (gspencergoog) [inline packages/flutter/test/material/raw_material_button_test.dart]: This should probably be a different year.

File: packages/flutter/test/widgets/app_test.dart [hunk 6]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/app_test.dart]: This should probably be a different year.

File: packages/flutter/test/widgets/async_lifecycle_test.dart [hunk 7]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/async_lifecycle_test.dart]: This should probably be a different year.

File: packages/flutter/test/widgets/physical_model_test.dart [hunk 8]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/physical_model_test.dart]: This should probably be a different year.

File: packages/flutter/test/widgets/sliver_constraints_test.dart [hunk 9]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/sliver_constraints_test.dart]: This should probably be a different year.

File: packages/flutter/test/widgets/test_border.dart [hunk 10]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/test_border.dart]: This should probably be a different year.

File: packages/flutter_tools/lib/src/base/user_messages.dart [hunk 11]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/base/user_messages.dart]: This should probably be a different year.

File: packages/flutter_tools/lib/src/build_runner/build_script_generator.dart [hunk 12]
```diff
@@ -1,4 +1,4 @@
-// Copyright 201 The Chromium Authors. All rights reserved.
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter_tools/lib/src/build_runner/build_script_generator.dart]: This should probably be a different year.

File: packages/flutter_tools/test/ios/simulators_test.dart [hunk 13]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter_tools/test/ios/simulators_test.dart]: This should probably be a different year.

File: packages/flutter_tools/test/src/pubspec_schema.dart [hunk 14]
```diff
@@ -1,3 +1,7 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (gspencergoog) [inline packages/flutter_tools/test/src/pubspec_schema.dart]: This should probably be a different year.

**Review by gspencergoog — CHANGES_REQUESTED** [review]
All the places where you add a year should be checked to make sure that they are the right year.

File: packages/flutter/test/material/raw_material_button_test.dart (hunk 4 above)
Comment (liyuqian) [reply packages/flutter/test/material/raw_material_button_test.dart]: Done. 2018.

File: packages/flutter/test/widgets/app_test.dart (hunk 6 above)
Comment (liyuqian) [reply packages/flutter/test/widgets/app_test.dart]: Done. 2018.

File: packages/flutter/test/widgets/physical_model_test.dart (hunk 8 above)
Comment (liyuqian) [reply packages/flutter/test/widgets/physical_model_test.dart]: Done. 2018.

File: packages/flutter/test/widgets/sliver_constraints_test.dart (hunk 9 above)
Comment (liyuqian) [reply packages/flutter/test/widgets/sliver_constraints_test.dart]: This is actuall 2019 😄

File: packages/flutter/test/widgets/test_border.dart (hunk 10 above)
Comment (liyuqian) [reply packages/flutter/test/widgets/test_border.dart]: It's 2019

File: packages/flutter_tools/lib/src/base/user_messages.dart (hunk 11 above)
Comment (liyuqian) [reply packages/flutter_tools/lib/src/base/user_messages.dart]: Done. 2018.

File: packages/flutter_tools/lib/src/build_runner/build_script_generator.dart (hunk 12 above)
Comment (liyuqian) [reply packages/flutter_tools/lib/src/build_runner/build_script_generator.dart]: It's 2019

File: packages/flutter_tools/test/ios/simulators_test.dart (hunk 13 above)
Comment (liyuqian) [reply packages/flutter_tools/test/ios/simulators_test.dart]: Done. 2017.

File: packages/flutter_tools/test/src/pubspec_schema.dart (hunk 14 above)
Comment (liyuqian) [reply packages/flutter_tools/test/src/pubspec_schema.dart]: Done. 2018

File: packages/flutter/test/widgets/async_lifecycle_test.dart (hunk 7 above)
Comment (liyuqian) [reply packages/flutter/test/widgets/async_lifecycle_test.dart]: Done. 2018

**Review by liyuqian — COMMENTED** [review]
Oops, sorry that I missed other files. Thanks for the remind!

**Review by gspencergoog — APPROVED** [review]

**Pull request merged**